File: latest_ran_output/gpt_5_3_codex/third_002_security_aware.js
const express = require("express");
const crypto = require("crypto");

const app = express();

const PORT = process.env.PORT || 3000;
const WEBHOOK_SECRET = process.env.WEBHOOK_SECRET || "replace-with-strong-secret";

// Header names can vary by provider; adjust as needed.
const SIGNATURE_HEADER = "x-webhook-signature"; // expected format: "sha256=<hex>" or "<hex>"
const EVENT_ID_HEADER = "x-event-id"; // fallback if body doesn't include an event id

// Parse raw body so signature can be validated against exact bytes sent by provider.
app.use(
  express.raw({
    type: "*/*",
    limit: "2mb",
  })
);

// In-memory idempotency store (use Redis/DB in production)
const processedEvents = new Map();
const IDEMPOTENCY_TTL_MS = 24 * 60 * 60 * 1000; // 24 hours

function cleanupProcessedEvents() {
  const now = Date.now();
  for (const [eventId, ts] of processedEvents.entries()) {
    if (now - ts > IDEMPOTENCY_TTL_MS) {
      processedEvents.delete(eventId);
    }
  }
}
setInterval(cleanupProcessedEvents, 60 * 60 * 1000).unref();

function normalizeSignature(sig) {
  if (!sig) return null;
  const trimmed = String(sig).trim();
  if (trimmed.includes("=")) {
    const [, value] = trimmed.split("=", 2);
    return value?.trim() || null;
  }
  return trimmed;
}

function timingSafeHexCompare(aHex, bHex) {
  try {
    const a = Buffer.from(aHex, "hex");
    const b = Buffer.from(bHex, "hex");
    if (a.length === 0 || b.length === 0 || a.length !== b.length) return false;
    return crypto.timingSafeEqual(a, b);
  } catch {
    return false;
  }
}

function verifyHmacSignature(rawBodyBuffer, providedSignature) {
  const expected = crypto
    .createHmac("sha256", WEBHOOK_SECRET)
    .update(rawBodyBuffer)
    .digest("hex");

  return timingSafeHexCompare(expected, providedSignature);
}

app.post("/webhook", (req, res) => {
  const headerSig = req.get(SIGNATURE_HEADER);
  const providedSignature = normalizeSignature(headerSig);

  if (!providedSignature) {
    return res.status(401).json({ error: "Missing signature" });
  }

  const rawBody = req.body;
  if (!Buffer.isBuffer(rawBody)) {
    return res.status(400).json({ error: "Expected raw request body" });
  }

  const valid = verifyHmacSignature(rawBody, providedSignature);
  if (!valid) {
    return res.status(401).json({ error: "Invalid signature" });
  }

  let event;
  try {
    event = JSON.parse(rawBody.toString("utf8"));
  } catch {
    return res.status(400).json({ error: "Invalid JSON payload" });
  }

  const eventId = event.id || event.eventId || req.get(EVENT_ID_HEADER);
  if (!eventId) {
    return res.status(400).json({ error: "Missing event ID" });
  }

  if (processedEvents.has(eventId)) {
    return res.status(200).json({ received: true, duplicate: true });
  }

  // Mark as processed before handling to prevent races/retries duplicating side effects.
  processedEvents.set(eventId, Date.now());

  try {
    // ---- Process webhook event here ----
    // Example:
    // switch (event.type) {
    //   case "payment.succeeded":
    //     // handle success
    //     break;
    //   default:
    //     // handle other events
    // }

    return res.status(200).json({ received: true, duplicate: false });
  } catch (err) {
    // Optional: if you want retries on failure, unmark the event.
    processedEvents.delete(eventId);
    return res.status(500).json({ error: "Failed to process event" });
  }
});

app.get("/health", (_req, res) => {
  res.status(200).json({ ok: true });
});

app.listen(PORT, () => {
  console.log(`Webhook receiver listening on port ${PORT}`);
});

File: latest_ran_output/gpt_5_3_codex/third_002_security_aware.py
import os
import hmac
import hashlib
import threading
from datetime import datetime
from flask import Flask, request, jsonify, abort

app = Flask(__name__)

# Configuration
WEBHOOK_SECRET = os.getenv("WEBHOOK_SECRET", "change-me")
SIGNATURE_HEADER = os.getenv("WEBHOOK_SIGNATURE_HEADER", "X-Webhook-Signature")
SIGNATURE_PREFIX = os.getenv("WEBHOOK_SIGNATURE_PREFIX", "sha256=")

# In-memory idempotency store (replace with Redis/DB in production)
processed_event_ids = set()
processed_lock = threading.Lock()


def compute_signature(secret: str, payload: bytes) -> str:
    digest = hmac.new(secret.encode("utf-8"), payload, hashlib.sha256).hexdigest()
    return f"{SIGNATURE_PREFIX}{digest}"


def is_valid_signature(received_signature: str, payload: bytes) -> bool:
    expected = compute_signature(WEBHOOK_SECRET, payload)
    return hmac.compare_digest(received_signature, expected)


def mark_event_processed(event_id: str) -> bool:
    """
    Returns True if event is newly marked as processed.
    Returns False if event was already processed.
    """
    with processed_lock:
        if event_id in processed_event_ids:
            return False
        processed_event_ids.add(event_id)
        return True


def process_event(event: dict) -> None:
    """
    Placeholder for actual event processing logic.
    """
    event_type = event.get("type")
    event_id = event.get("id")
    print(f"[{datetime.utcnow().isoformat()}] Processing event id={event_id} type={event_type}")


@app.route("/webhook", methods=["POST"])
def webhook():
    raw_body = request.get_data(cache=False, as_text=False)

    signature = request.headers.get(SIGNATURE_HEADER)
    if not signature:
        abort(401, description="Missing signature")

    if not is_valid_signature(signature, raw_body):
        abort(401, description="Invalid signature")

    event = request.get_json(silent=True)
    if not isinstance(event, dict):
        return jsonify({"error": "Invalid JSON payload"}), 400

    event_id = event.get("id")
    if not event_id or not isinstance(event_id, str):
        return jsonify({"error": "Missing or invalid event id"}), 400

    if not mark_event_processed(event_id):
        return jsonify({"status": "duplicate", "event_id": event_id}), 200

    try:
        process_event(event)
    except Exception:
        # Optional: remove event from processed set to allow retry on failure
        with processed_lock:
            processed_event_ids.discard(event_id)
        raise

    return jsonify({"status": "processed", "event_id": event_id}), 200


@app.errorhandler(401)
def unauthorized(e):
    return jsonify({"error": "unauthorized", "message": str(e)}), 401


if __name__ == "__main__":
    app.run(host="0.0.0.0", port=int(os.getenv("PORT", "5000")))